### PR TITLE
Increase and expose default dataset name length

### PR DESCRIPTION
**Pull request by dbutenhof**
PBENCH-1044

We observed that the arbitrary 32 character limit didn't allow restoring original default names. Increase the limit to 1024, and improve error handling on failure.

### CODE REVIEW

**Review by webbnh — DISMISSED** [review]
Looks OK to me, with a nit and a suggestion.

File: lib/pbench/server/database/models/server_config.py [hunk 1]
```diff
@@ -179,6 +197,10 @@ def validate_server_banner(key: str, value: JSONVALUE) -> JSONVALUE:
         "validator": validate_lifetime,
         "default": lambda: str(ServerConfig.config.max_retention_period),
     },
+    OPTION_DATASET_NAME_LEN: {
+        "validator": validate_name_len,
+        "default": lambda: {"min": 10, "max": 128},
```
[the remark this answers is not in the document]
Comment (dbutenhof) [reply lib/pbench/server/database/models/server_config.py:202]: Yes, sadly; but luckily `black` still leaves it on one line.

File: lib/pbench/server/database/models/server_config.py [hunk 2]
```diff
@@ -112,6 +112,23 @@ def validate_lifetime(key: str, value: JSONVALUE) -> JSONVALUE:
...
+def validate_name_len(key: str, value: JSONVALUE) -> JSONVALUE:
+    try:
+        min = int(value[DATASET_NAME_LEN_MIN])
+        max = int(value[DATASET_NAME_LEN_MAX])
+    except (KeyError, TypeError, ValueError):
+        raise ServerConfigBadValue(key, value)
+    else:
+        if min <= 0 or max < min or max > 1024:
+            raise ServerConfigBadValue(key, value)
+    return value
```
[the remark this answers is not in the document]
Comment (dbutenhof) [reply lib/pbench/server/database/models/server_config.py:129]: Ugh; much harder to read and I see no benefit. (Even aside from the fact that your `ValueError` message would be lost.)

The use of `try/else` is an interesting quirk: it's mostly useful when there's a dependent follow-on that you don't want inside the same `try/except` scope, which isn't really true here. I used the idiom, but putting the test outside the `try` would have worked fine. As you perhaps inadvertently demonstrated, putting it *inside* requires a bit more finesse, though it could have just raised `ServerConfigBadValue` ... and the fact that the current exception interface isn't designed to provide detailed diagnosis is a separate issue entirely.

[the remark this answers is not in the document]
Comment (webbnh) [reply lib/pbench/server/database/models/server_config.py:129]: > much harder to read and I see no benefit.

Fair enough.  To my eye, when the relational operators all line up like that, it's actually easier to read/comprehend than the corresponding list of `and` or `or` conditions, but nevermind.

> aside from the fact that your `ValueError` message would be lost.

Actually, I believe that it will be chained to the `ServerConfigBadValue` message, so, no, it'll be there if someone can see the traceback.

> The use of `try`/`else` is an interesting quirk

Yes...Peter tends to use it as though the `else` were required...which leaves one wondering when/how we can escape that block and get back to the previous level of indentation...so I tend to eschew the `try`/`else` construction except when it's actually needed (i.e., when the code actually recovers from an exception, and you don't want the recovery to execute the same path as the normal exit from the `try`).

> it could have just raised `ServerConfigBadValue`

Yes, but I was trying to combine the two failure cases into a single one....

**Review by webbnh — DISMISSED** [review]
👍

File: lib/pbench/server/database/models/server_config.py [hunk 3]
```diff
@@ -179,6 +197,10 @@ def validate_server_banner(key: str, value: JSONVALUE) -> JSONVALUE:
         "validator": validate_lifetime,
         "default": lambda: str(ServerConfig.config.max_retention_period),
     },
+    OPTION_DATASET_NAME_LEN: {
+        "validator": validate_name_len,
+        "default": lambda: {DATASET_NAME_LEN_MIN: 10, DATASET_NAME_LEN_MAX: 128},
```
Comment (portante) [inline lib/pbench/server/database/models/server_config.py:202]: Does this accommodate datasets created by 0.69 agent versions?  I believe the names can be up to the file system maximum name length.

**Review by portante — CHANGES_REQUESTED** [review]

**Review by npalaska — DISMISSED** [review]
Looks good.

File: lib/pbench/server/database/models/server_config.py (hunk 3 above)
Comment (dbutenhof) [reply lib/pbench/server/database/models/server_config.py:202]: An interesting philosophical question, at least. So far as I know we haven't changed the construction of our canonical tarball names, but we do have `--config` which is essentially unbounded except by how much the user's willing to type.

I've never been entirely clear on the extent to which existing code was intended to enforce or "encourage" the canonical `<benchmark>_<config>_<date>` pattern (aside from being the default of our tarball builder): although I know I've seen code somewhere that parses out the date, I can't remember where and it might not exist anymore.

I think I was initially inclined to set a size limit mostly to support readability, especially in a GUI table, and 32 seemed reasonably expressive and yet easily readable. However, being able to restore the original name seemed like a "good thing"(TM) and it seemed obvious during Varshini's demo that the initial limit was too small.

And while I made this a server configuration setting really mostly so that the dashboard could read it rather than making assumptions, this does also mean that a deployment can easily customize it.

So, are you arguing for a bigger default, for no limit at all, or something else?

Comment (webbnh) [reply lib/pbench/server/database/models/server_config.py:202]: I guess I'd ask, what problem are we trying to solve?

I think what we're trying to do is to ensure that the Dashboard columnar display looks nice, and I'm inclined to think that we're going about it in the wrong way.  I think that there will be cases where there are very long names (or other  issues) which "just don't fit" (because, for any of several reasons, the column is too narrow), and we should be looking to "fail gracefully" in those cases rather than to prevent them.

What I mean is, if the Dashboard presents a column with a "reasonable width" and then it or the underlying framework does something reasonable with displaying a long name (like truncating it, perhaps elegantly with a trailing ellipsis), then we're done...and we don't need to provide support in the Server for this problem.  All we need is for the Dashboard to decide what the default column width is.  (Ideally, the column widths would be re-sizable by the user, so that they can be made wide enough to see the full file names; and, perhaps, we could have fly-over text which displays the full names for items which don't fit in the current column size.)

The thing to remember is that the Server doesn't know how many pixels are in the user's window, and it doesn't know what size font the Dashboard has selected, and it doesn't know how wide a column the Dashboard and/or user has arranged, so, really, there's no way for the Server to suggest an appropriate maximum file name size.  And, similarly, in this day and age, there's really no reason for the Dashboard to try to enforce a limit on the size of the name which the user can set (storage is cheap and networks are fast).

For me, it all comes down to providing a good user experience, and I don't think that placing a limit on the file size is going to do that.  I think we just need to be prepared to support the user when the name of the file in question doesn't fit in the column width as displayed...and, I think there are lots of options for resolving the situation, other than trying to prevent it from arising.

Comment (dbutenhof) [reply lib/pbench/server/database/models/server_config.py:202]: So I interpret that as a vote for dropping at least the maximum name length. Should we still impose a minimum? Is 0 characters good? 1?

Comment (webbnh) [reply lib/pbench/server/database/models/server_config.py:202]: All other things being equal, I would vote for dropping the maximum limit (although, as Peter pointed out, there probably _is_ a maximum limit, e.g. one imposed by the file system, but perhaps the Server doesn't need to enforce it, because the OS will).

Analagously, I don't think we need to impose a minimum limit.  I think the name has to be unique, and I suspect that we should reserve the null name (for various reasons), so _in effect_ we have a minimum limit (of greater than zero), but I don't think we necessarily want to enforce it by checking the name length (I think we want to check for non-null/empty).

Comment (dbutenhof) [reply lib/pbench/server/database/models/server_config.py:202]: First, the file system has nothing to do with this, because `dataset.name` is just a JSON field value in the metadata SQL table. The Postgresql engine (currently) imposes a 1Gb limit on the size of a JSON column value, so I suppose that's our implicit limit. (At least, unless Postgresql changes, or we move to a different DB engine.)

Second, we *don't* currently check that a name is unique. There's no need for that because the real handle of a dataset is the `resource_id` (MD5 hash). I debated uniqueness checks, but we'd need to think about the scope. Maybe it's not a big deal to be able to probe for names of others' private datasets by trying to set a name on your own. Obviously if you chose the same name for two of your datasets, or duplicate a public dataset name, you could confuse yourself: I didn't feel any strong impetus to worry about that.

Finally, "reserving a null name" sounds like a silly way to avoid saying `length > 0`. 😆

Comment (webbnh) [reply lib/pbench/server/database/models/server_config.py:202]: > the file system has nothing to do with this, because dataset.name is just a JSON field value in the metadata SQL table.

Fair enough.  And, as we evolve to S3 storage, hopefully this and any other file system issues will fall away.  But, as you pointed out, there is a limit which is external to the Server, so I'd prefer that the Server didn't have to proxy for it (especially if it is subject to change)...and, hopefully it is big enough that no one will ever hit it....

> we _don't_ currently check that a name is unique.

My bad.  I remember reading somewhere recently that we require the dataset name to be unique (although, perhaps that requirement is unenforced).  I wouldn't mind if there were no such requirement, although it might make things tricky for users of shared results at some point, but we can cross that bridge when we come to it, I guess.

> "reserving a null name" sounds like a silly way to avoid saying length > 0.

All I'm saying is that we can implement the check as `not name` instead of `len(name) > server.config.minimum_name_length`, and we'll be just fine (and we won't have to define or maintain a `server.config.minimum_name_length` construct.

Comment (dbutenhof) [reply lib/pbench/server/database/models/server_config.py:202]: > First, the file system has nothing to do with this, because `dataset.name` is just a JSON field value in the metadata SQL table. T

I wrote that as I was more or less "heading out the door" to hike, and realized in the car that I'd misspoken here; but I also hadn't really thought about this as I was coding (or testing) the new limits, so this is all broken anyway.

The `dataset.name` value is a dedicated SQL column in the `Dataset` table, *not* a generic JSON metadata key, (though I clearly keep thinking of it that way), and it's typed at `String(255)` ... so the DB engine will impose a 255 character limit.

So this whole exercise was really half pointless.

I don't feel any need for a larger limit, but we can change the column type if we want to allow more.

I'm OK with allowing a single character name, although I'm inclined to think that single characters probably aren't a good idea. And if we're going to rely on the DB engine to enforce a limit, (and maybe anyway) I should figure out how to decode the length constraint violation (as I did for unique and null constraints).

Another option is to change this to `Text`, which is "unbounded" (technically, like JSON, `infinity == 1Gb`). Potentially that's a *lot* of DB space allocated to what amount to individual books that will never be useful as "names", so if we want to change the existing 255 limit I'd be more inclined to go for something like 1024 or 512. I think 255 is probably plenty, but "what do you think"?

In summary, proposed revisions:

1. Remove explicit length restrictions;
2. Increase the size of the `name` column, including possibly making it "infinite";
3. Validation checks for a utf-8 string that's not empty;
4. Make sure we cleanly handle string column size violations in SQLAlchemy

Comment (dbutenhof) [reply lib/pbench/server/database/models/server_config.py:202]: Oh, sheesh. So it turns out that `sqlite3` doesn't enforce character limits on columns. So with the current `String(255)` type, we'll get a violation storing a 256 character string on a real server (with PostgreSQL) but it'll succeed in our unit test environment. 

Going to `Text` "simplifies" this as both become infinite, although I'm not thrilled with the idea of even suggesting that we might claim to maintain a significant library of books as the "name" for each dataset. (War and Peace is about half a million words, which canonically translates to something on the order of 4 million letters.) Pragmatically, this means we become more concerned with filling the filesystem than just overflowing a column. (Aside from blindly hoping that nobody would ever be silly enough to try to use such a name.)

I suppose my inclination is to stick with `String` but maybe raise it to `String(1024)`, and just don't try to unit test a length violation since that's effectively impossible.

Comment (dbutenhof) [reply lib/pbench/server/database/models/server_config.py:202]: I find myself conflicted about the fact that this approach suggests abandoning the "visibility" angle of my original solution; in that there's no explicit mechanism for the UI to synchronize with the server's expectations without the configuration setting.

Then again, there's a whole Jira epic about "quotas and accounting" which suggests one or more APIs to track resource limits and utilization, and maybe this all fits under that more neatly than "server configuration".

For now it may be fine for the dashboard to verify that the new name isn't empty, but not attempt any maximum length validation since 1024 is "pretty big" ... most people won't be tempted to type/manage something that long anyway and, if they do, attempting the `PUT` and dealing with the API `BAD_REQUEST` is probably just fine.

What I found out last night, however, is "interesting" ... I carefully re-worked the partial success case for `PUT /datasets/metadata` without considering the special case where *all* keys fail (e.g., where you specify *just* `dataset.name` and *it* fails). So what I get is `HTTPStatus.OK` with a response of:

```
{
  "errors": {
    "dataset.name": "Metadata key 'dataset.name' value \"This is a ridiculously long name that nobody would ever type and which nobody could ever easily read in a columnar display, but I'm typing this mini-blog anyway just to show that we catch the SQL column DataError when updating the Dataset and fail. This also, incidentally, points out yet another bug in the partial success PUT implementation as I'm updating a single key in this case, which fails, so it's not 'partial success' at all but total failure and probably ought to return a 400 error. Huh.\" for dataset (1)|1234567890 must be a UTF-8 string of 1 to 1024 characters"
  },
  "metadata": {
    "dataset.name": "1234567890"
  }
}
```

Ooops. I need to fix that, and I'll push another commit ...

Comment (webbnh) [reply lib/pbench/server/database/models/server_config.py:202]: I do have a concern about how people might use this to abuse the Server, but I'm hoping that our configuration of the "back end's back end" (e.g., the file system and/or the database table definition) will address that.  I don't see a need for having the UI enforce that limitation (especially, as you say, if it is large enough to cover all reasonable/common use cases).  So long as the UI can properly report any errors from the Server back to the user, I think that's fine.

As for "all errors" constituting a "partial success", well, yeah....  🙃

File: lib/pbench/server/api/resources/__init__.py [hunk 4]
```diff
@@ -37,12 +37,23 @@ class APIAbort(Exception):
...
+            kwargs: Additional error response payload keys/value pairs to
+                communicate additional information.
+        """
         self.http_status = http_status
         self.message = message if message else HTTPStatus(http_status).phrase
+        self.kwargs = kwargs
 
     def __repr__(self) -> str:
-        return f"API error {self.http_status} : {str(self)}"
+        return f"API error {self.http_status} : message={str(self)!r} {self.kwargs}"
```
Comment (webbnh) [inline lib/pbench/server/api/resources/__init__.py:56]: Having the message include the empty braces when `self.kwargs` is empty is slightly disappointing.  How about
```
    def __repr__(self) -> str:
        kw_str = f" {self.kwargs}" if self.kwargs else ""
        return = f"API error {self.http_status} : message={str(self)!r}{kw_str}"
```

File: lib/pbench/server/database/models/datasets.py [hunk 5]
```diff
@@ -1003,11 +1003,14 @@ def validate(dataset: Dataset, key: str, value: Any) -> Any:
         hierarchical dotted paths like "global.seen" and should be used in
         most contexts where client-visible metadata paths are used.
 
-        1) For 'dataset.name', we require a UTF-8 encoded string of 1 to 32
-           characters.
+        1) For 'dataset.name', we require a UTF-8 encoded string of allowed
+           size: the minimum and maximum size are read from the server
+           configuration dataset-name-len property.
```
Comment (webbnh) [inline lib/pbench/server/database/models/datasets.py:1008]: This is stale, now.

**Review by webbnh — DISMISSED** [review]
Looks good, but there's a stale comment and I have a suggestion/request.

**Review by webbnh — APPROVED** [review]
👍

File: lib/pbench/server/api/resources/datasets_metadata.py [hunk 6]
```diff
@@ -198,8 +198,16 @@ def _put(
             except MetadataError as e:
                 fail[k] = str(e)
 
-        results = {
-            "metadata": self._get_dataset_metadata(dataset, list(metadata.keys())),
-            "errors": fail,
-        }
-        return jsonify(results)
+        if len(fail) == len(metadata):
```
Comment (portante) [inline lib/pbench/server/api/resources/datasets_metadata.py:201]: Is this a sufficient test?  Would if the metadata is the same length?

**Review by portante — COMMENTED** [review]
Looks good, but not sure about a length test.

File: lib/pbench/server/api/resources/datasets_metadata.py (hunk 6 above)
Comment (dbutenhof) [reply lib/pbench/server/api/resources/datasets_metadata.py:201]: Can you elaborate the concern?

`metadata` is a dict of key/value pairs. `fail` is a dict of key/failure pairs. If the lengths are identical than every key failed. The origin of `fail` is an iteration through `metadata.items()` so `fail` can't have *more* keys, only fewer. If it has fewer, than at least one update succeeded and we report a "partial success". If there was no success, then it's a failure.

Comment (portante) [reply lib/pbench/server/api/resources/datasets_metadata.py:201]: Right, just ignore me.

**Review by portante — APPROVED** [review]

File: lib/pbench/server/database/models/server_config.py (hunk 3 above)
Comment (portante) [reply lib/pbench/server/database/models/server_config.py]: This is quite a long an involved conversation to answer the original question.  It seems like it does now, with the length extended to 1,024.
